google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java

```diff
@@ -394,8 +394,15 @@ protected static String getAppEngineProjectId() {
       String serviceAccountName = (String) method.invoke(appIdentityService);
       int indexOfAtSign = serviceAccountName.indexOf('@');
       return serviceAccountName.substring(0, indexOfAtSign);
+    } catch (ClassNotFoundException exception) {
+      if (System.getProperty("com.google.appengine.runtime.version") != null) {
+        // Could not resolve appengine classes under GAE environment.
+        throw new RuntimeException("Unable to resolve appengine-sdk classes. "
+            + "For more details see "
```

Comment:
Sorry about one last request, but I didn't think of it until now and I think it will help even more. I think we should add something about the App Engine runtime being detected. Something like `"Google App Engine runtime detected (the environment variable \"com.google.appengine.runtime.version\" is set), but unable to resolve appengine-sdk classes. "`